Rework dev docs (#1520)
tektoncd/dashboard: docs/dev/README.md

# Tekton Dashboard - Dev docs

- Running tests

> - Troubleshooting

Reviewer: These seem a little odd here, can you remove them? They're included in the table of contents below anyway.

Author: fixed

This guide explains how to build, deploy and test the Tekton Dashboard. It covers the following topics:

- [Before you begin](#before-you-begin)
- [Pre-requisites](#pre-requisites)
- [Checkout your fork](#checkout-your-fork)
- [Build the frontend](#build-the-frontend)
- [Build the backend](#build-the-backend)
  - [Command line arguments](#command-line-arguments)
- [Build YAML manifests](#build-yaml-manifests)
- [Build and deploy with kustomize and ko](#build-and-deploy-with-kustomize-and-ko)
  - [Build docker image for another GOARCH](#build-docker-image-for-another-goarch)
- [Development server](#development-server)
- [Run backend tests](#run-backend-tests)
- [Run frontend tests](#run-backend-tests)
  - [Unit tests](#unit-tests)
  - [Linter](#linter)
- [i18n](#i18n)
- [Storybook](#storybook)
- [Troubleshooting](#troubleshooting)
- [Next steps](#next-steps)

## Before you begin

Many of the instructions here replicate what's in the [tektoncd/pipeline development guide](https://github.com/tektoncd/pipeline/blob/master/DEVELOPMENT.md), with a couple of caveats currently.

> - This project has not yet been tested with GKE

Reviewer: The dogfooding and robocat clusters are on GKE. I think we can just remove this line.

Author: was not sure, copied from the old docs

## Pre-requisites

> In order to run the Tekton Dashboard, please make sure the requirements are met in [the install doc](../install.md) are met.

Reviewer: ```suggestion
In order to run the Tekton Dashboard, please make sure the requirements in [the install doc](../install.md) are met.
```

> You will also need the following tool in order to build the Dashboard locally and deploying it:

Reviewer: ```suggestion
You will also need the following tools in order to build the Dashboard locally and deploy it:
```

> 1. [`go`](https://golang.org/doc/install): The language Tekton Dashboard is built in

Reviewer: ```suggestion
1. [`go`](https://golang.org/doc/install): The language the Tekton Dashboard backend is built in
```

1. [`git`](https://help.github.com/articles/set-up-git/): For source control
1. [Node.js & npm](https://nodejs.org/): For building and running the frontend locally. See `engines` in [package.json](../../package.json) for versions used. _Node.js 10.x is recommended_

> 1. [`ko`](https://github.com/google/ko): For development. `ko` version v0.1 or higher is required for `dashboard` to work correctly

Reviewer: soon to be 0.5.1 if we go ahead with https://github.com/tektoncd/dashboard/pull/1527

Author: i changed it to 0.5.0 until we switch to 0.5.1

1. [`kubectl`](https://kubernetes.io/docs/tasks/tools/install-kubectl/): For interacting with your kube cluster
1. [`kustomize`](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/INSTALL.md): For building the Dashboard manifests. You need a recent version - v3.5.4 is recommended
See [here](https://github.com/kubernetes-sigs/kustomize/blob/master/docs/INSTALL.md#try-go) - `GO111MODULE=on go install sigs.k8s.io/kustomize/kustomize/v3` works correctly

Note that there exists a bug in certain versions of `kubectl` whereby the `auth` field is missing from created secrets. Known good versions we've tested are __1.15.0__ and __1.16.2__.

## Checkout your fork

To check out this repository:

1. Create your own [fork of this repo](https://help.github.com/articles/fork-a-repo/)
1. Clone it to your machine:

```bash

> mkdir -p ${GOPATH}/src/github.com/tektoncd

Reviewer: Does it need to be on GOPATH anymore since the switch to modules? I haven't run into any issues with it outside GOPATH

Edit: I see the note below. I think we should remove references to GOPATH here to avoid confusion.

Author: fixed

| Argument | Description | Type | Default value |
|---|---|---|---|
| `--csrf-secure-cookie` | Enable or disable Secure attribute on the CSRF cookie | `bool` | `true` |
| `--help` | Print help message and exit | `bool` | `false` |
| `--kube-config` | Path to kube config file, uses in cluster config if empty | `string` | `""` |
| `--pipelines-namespace` | Namespace where Tekton pipelines is installed (assumes same namespace as dashboard if not set) | `string` | `""` |
| `--triggers-namespace` | Namespace where Tekton triggers is installed (assumes same namespace as dashboard if not set) | `string` | `""` |
| `--port` | Dashboard port number | `int` | `8080` |
| `--read-only` | Enable or disable read only mode | `bool` | `false` |
| `--web-dir` | Dashboard web resources directory | `string` | `""` |
| `--logout-url` | If set, enables logout on the frontend and binds the logout button to this url | `string` | `""` |
| `--namespace` | If set, limits the scope of resources watched to this namespace only | `string` | `""` |
| `--log-level` | Minimum log level output by the logger | `string` | `"info"` |
| `--log-format` | Format for log output (json or console) | `string` | `"json"` |

> Run `dashboard --help` to show the supported command line arguments and their default value directly from the `dashboard` binary.

Reviewer: ```suggestion
Run `dashboard --help` to show the supported command line arguments and their default values directly from the `dashboard` binary.
```

**Important note:** using `--namespace` ensures that the dashboard is watching resources in the namespace specified (and drives the frontend).

> It doesn't limit actions that can be performed to this namespace only though. It's important that this flag is used AND that rbac rules are setup accordingly.

Reviewer: ```suggestion
It doesn't limit actions that can be performed to this namespace only though. It's important that this flag is used **and** that RBAC rules are setup accordingly.
```

## Build YAML manifests

Manifests to be deployed come in multiple flavors and are generated using `kustomize`.

The `kustomize` overlays can be found in the overlays folder, they cover the target platform (plain kube or Openshift) as well as the Dashboard variation (read-only or read-write).

> Use one of the following command to get the manifests that suits your need:

Reviewer: ```suggestion
Use one of the following commands to get the manifest that suits your needs:
```

```bash
# Plain Kube
kustomize build overlays/dev
kustomize build --load_restrictor none overlays/dev-locked-down

# OpenShift
kustomize build overlays/dev-openshift --load_restrictor none
kustomize build overlays/dev-openshift-locked-down --load_restrictor none
```

We also provide a version for working with OpenShift image streams:
```bash
kustomize build overlays/dev-openshift-imagestream --load_restrictor none
```

## Build and deploy with kustomize and ko

> `kustomize` and `ko` can be used together to build the yaml manifests, build and push the dahsboard docker image, and deploy the results in a cluster.

Reviewer: ```suggestion
`kustomize` and `ko` can be used together to build the YAML manifests, build and push the Dashboard docker image, and deploy the results in a cluster.
```

> Note that it doesn't cover the frontend, make sure you [built the frontend](#build-the-frontend) before.

Reviewer: ```suggestion
Note that it doesn't cover the frontend, make sure you [build the frontend](#build-the-frontend) before.
```

Before `ko` can build and push the dashboard docker image you will have to define the `KO_DOCKER_REPO` environment variable.

```bash

> export KO_DOCKER_REPO='docker.io/myusername'

Reviewer: alternatively `ko.local` to avoid dealing with auth / network issues for an external repo

Author: neveer used `ko.local` this is the default for `ko` if `KO_DOCKER_REPO` is not defined ?

Reviewer: no it's the same as doing `ko --local` or `ko -L` https://github.com/google/ko#usage
I use it most of the time, only using docker hub when I need to. Even though our image is small I've found docker hub can often hang for a while before uploading and `ko.local` avoids that problem.

Author: i see, fixed

```

> Depending on the dashboard flavour you want to deploy, use one of the following command:

Reviewer: ```suggestion
Depending on the dashboard flavour you want to deploy, use one of the following commands:
```

```

To run unit tests with `-race`:

```bash
# CGO_ENABLED=1 is needed for -race on go test
CGO_ENABLED=1 NAMESPACE=default go test -race -v ./...
```

## Run frontend tests

### Unit tests

Run `npm test` to execute the unit tests via [Jest](https://jestjs.io/) in interactive watch mode. This also generates a code coverage report by default.

Coverage threshold is set to 90%, if it falls below the threshold the test script will fail.

Tests are defined in `*.test.js` files alongside the code under test.

### Linter

Run `npm run lint` to execute the linter (eslint + prettier). This will ensure code follows the conventions and standards used by the project.

Run `npm run lint:fix` to automatically fix a number of types of problem including code style.

## i18n

This project uses `react-intl` for internationalization, and provides a script to automatically extract messages to bundles for translation. Run `npm run i18n:extract` from the root of the project to update the message bundles which can be found in `src/nls/`.

**Note:** `src/nls/messages_en.json` should **NOT** be edited manually, instead edit the defaultMessage in the code and re-run the script.

## Storybook

> Run `npm run storybook` to start [storybook](https://storybook.js.org/) in development mode. Navigate to [`http://localhost:4000/`](http://localhost:4000/) in your browser. The app will automatically hot-reload any changes to the source files, including CSS.

Reviewer: ```suggestion
Run `npm run storybook` to start [storybook](https://storybook.js.org/) in development mode. It should automatically open [`http://localhost:5000/`](http://localhost:5000/) in your browser. The app will automatically hot-reload any changes to the source files, including CSS.
```
